holmes/plugins/toolsets/datadog/toolset_datadog_logs.py

```diff
@@ -0,0 +1,351 @@
+
+BASE_RETRY_DELAY = 2.0  # Initial fallback delay if datadog does not return a reset_time
+MAX_RETRY_DELAY = 60.0  # Max delay if datadog does not return a reset_time
+MAX_RETRY_COUNT_ON_RATE_LIMIT = 5
+
+
+class DataDogLabelsMapping(BaseModel):
+    pod: str = "pod_name"
+    namespace: str = "kube_namespace"
+
+
+class DataDogStorageTier(str, Enum):
+    INDEXES = "indexes"
+    ONLINE_ARCHIVES = "online-archives"
+    FLEX = "flex"
+
+
+DEFAULT_STORAGE_TIERS = [DataDogStorageTier.INDEXES]
+
+
+class DatadogConfig(BaseModel):
+    dd_api_key: str
+    dd_app_key: str
+    indexes: list[str] = ["*"]
+    site_api_url: str  # e.g. https://api.us3.datadoghq.com. c.f. https://docs.datadoghq.com/getting_started/site/
+    # Ordered list of storage tiers. Works as fallback. Subsequent tiers are queried only if the previous tier yielded no result
+    storage_tiers: list[DataDogStorageTier] = DEFAULT_STORAGE_TIERS
+    labels: DataDogLabelsMapping = DataDogLabelsMapping()
+    page_size: int = 300
+    default_limit: int = 1000
+    request_timeout: int = 60
+
+
+class DataDogRequestError(Exception):
+    payload: dict
+    status_code: int
+    response_text: str
+
+    def __init__(self, payload: dict, status_code: int, response_text: str):
+        super().__init__(f"HTTP error: {status_code} - {response_text}")
+        self.payload = payload
+        self.status_code = status_code
+        self.response_text = response_text
+
+
+class DataDogRequest429Error(DataDogRequestError):
+    reset_time: Optional[int]
+
+    def __init__(
+        self,
+        payload: dict,
+        status_code: int,
+        response_text: str,
+        reset_time: Optional[int],
+    ):
+        super().__init__(
+            payload=payload, status_code=status_code, response_text=response_text
+        )
+        self.reset_time = reset_time
+
+
+def _extract_cursor_for_next_paginated_api_call(data: dict) -> Optional[str]:
+    return data.get("meta", {}).get("page", {}).get("after", None)
+
+
+def _execute_logs_query(
+    url: str, headers: dict, payload: dict, timeout: int
+) -> tuple[list[dict], Optional[str]]:
+    attempts_count = 0
+
+    while attempts_count < MAX_RETRY_COUNT_ON_RATE_LIMIT:
+        attempts_count += 1
+        response = requests.post(url, headers=headers, json=payload, timeout=timeout)
+
+        if response.status_code == 200:
+            data = response.json()
+            cursor = _extract_cursor_for_next_paginated_api_call(data)
+
+            logs = data.get("data", [])
+            return logs, cursor
+
+        elif response.status_code == 429:
+            wait_time = min(
+                BASE_RETRY_DELAY * (2 ** (attempts_count - 1)), MAX_RETRY_DELAY
+            )
+
+            reset_time_header = response.headers.get("X-RateLimit-Reset")
+            if reset_time_header:
+                try:
+                    reset_time = int(reset_time_header)
+                    wait_time = max(0, reset_time) + 0.1
```

Comment:
Product wise do we want to be able to limit this period? 
what is the max time that throttle can we get here? how it can effect the user expirience?